src/server/journal/streamer.h

```diff
@@ -62,9 +65,59 @@ class JournalStreamer {
   bool IsStalled() const;
 
   journal::Journal* journal_;
-  std::vector<uint8_t> pending_buf_;
-  size_t in_flight_bytes_ = 0, total_sent_ = 0;
 
+  class PendingBuf {
+   public:
+    struct Buf {
+      size_t mem_size = 0;
+      absl::InlinedVector<std::string, 8> buf;
+
+      static constexpr size_t max_buf_size = 1024;  // depends on UIO_MAXIOV
+    };
+
+    PendingBuf() : bufs_(1) {
+    }
+
+    bool empty() const {
+      return std::all_of(bufs_.begin(), bufs_.end(), [](const auto& b) { return b.buf.empty(); });
+    }
+
+    void push(std::string str) {
+      DCHECK(!bufs_.empty());
+      if (bufs_.back().buf.size() == Buf::max_buf_size) {
+        bufs_.emplace_back();
+      }
+      auto& fron_buf = bufs_.back();
+
+      fron_buf.mem_size += str.size();
+      fron_buf.buf.push_back(std::move(str));
+    }
+
+    // should be called to get the next buffer for sending
+    const Buf& PrepareNext() {
+      if (bufs_.size() == 1) {
+        bufs_.emplace_back();
+      }
```

Comment:
I feel like it would be more robust to handle this case in `Pop()` below rather than here.
I.e. do in `Pop()` something like: `bufs_.pop_front(); if empty { bufs.emplace_back(); }`

Reply:
we can add new items before pop, but we shouldn't update the one that is already sent. So I can do it.